packages/cosmic-proto/src/codec-helpers.ts

```diff
@@ -0,0 +1,158 @@
+import type { BinaryReader, BinaryWriter, JsonSafe } from './codegen/index.js';
+import type { MessageBody, TypedJson } from './helpers.js';
+
+const { freeze } = Object;
+
+export type ProtoMsg<TU = string> = {
+  readonly typeUrl: TU;
+  readonly value: Uint8Array;
+};
+
+export interface EncodeObject<TU = string> {
+  readonly typeUrl: TU;
+  readonly value: MessageBody<TU>;
+}
+
+export interface TypedAmino<TU = string, MT = MessageBody<TU>> {
+  readonly type: TU;
+  readonly value: MT;
+}
+
+const extractTypeUrlAndValue = (input: any) => {
+  if ('@type' in input) {
+    // TypedJson
+    const { '@type': typeUrl, ...value } = input;
+    return { typeUrl, value };
+  } else if ('typeUrl' in input) {
+    // ProtoMsg (value: Uint8Array) or EncodeObject (value: MessageBody)
+    const { typeUrl, value } = input;
+    return { typeUrl, value };
```

Comment:
```suggestion
    return { typeUrl, value: {...value} };
```

To avoid keeping the same object, and potentially have `fromTyped` mutate it.

Reply:
Thanks!